### ISSUE TITLE
Livepeer Zero-Downtime Deployment

### ISSUE THREAD

**Issue by leszko** [livepeer/go-livepeer#2129]
**Is your feature request related to a problem? Please describe.**
We need to prepare a guide on how to deploy Livepeer (Orchestrator) which supports upgrade or config change without losing the streams.

Related GH Issues:
- https://github.com/livepeer/go-livepeer/issues/2047
- https://github.com/livepeer/go-livepeer/issues/2058

**Describe the solution you'd like**
There should be a possibility of restarting Livepeer without losing streams.

[Comment](https://github.com/livepeer/go-livepeer/issues/2047#issuecomment-985065858) from @hthillman 
> Provide a guide for restarting livepeer if you need to change your configuration without losing any streams that are currently coming in by spinning up a node with the new configuration, moving streams over to that new node and then shutting down the existing node. Something similar to this Lifecycle - Lotus Docs 1 (but hopefully simpler) along with some technical changes to make it easier to do

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by leszko** [comment]
@yondonfu @hthillman I've spent some thinking about how we should organize and here are my thoughts. Let me know what you think.

**1. Documentation**

I think we should create a new section in [Livepeer Org, Video Miners, How-to Guides](https://livepeer.org/docs/video-miners/how-to-guides) called "Zero Downtime Deployment". It will be complementary to [Getting Started: Activation](https://livepeer.org/docs/video-miners/getting-started/activation) where we describe the most basic setup. The new section is intended to describe how to deploy Livepeer in a way to allow configuration update and Livepeer version upgrade without losing the current streams.

**2. Solution**

I think we should describe a solution similar to what we do in livepeer.com, but probably not mention anything about Kubernetes for now. Livepeer should always run behind a proxy, for the simplest on-premises setup we can describe `socat` or `nginx` (in livepeer.com we use `traefik`, but it fits better Docker/Kubernetes environments). Then, for the Livepeer update/ugprade, a user does the following:
1. Start a new Livepeer instance
2. Switch the proxy redirect
3. Stop the old instance

So, a classic Blue-Green Deployment.

One caveat of this approach is that if there is any segment in the process of transcoding, then it may be lost. I guess it can be solved with the use of Google Cloud Storage or S3 instead of the deafult memory session. However, it'd address this issue separately, because it's a problem on its own, which touches both Zero Downtime Deployment and High Availabily.

**3. Alternative solution**

One alternative would be to bake a redirect mechanims into Livepeer. It was already discussed in Discord [here](https://discord.com/channels/423160867534929930/445623681251672064/704779780897243228) and is partly implemented, becase a new O address can be returned [here](https://github.com/livepeer/go-livepeer/blob/bf1c0259f7a36b1838f00a5af5967c5c107f0162/net/lp_rpc.proto#L106). Such solution could work as follows:

1. Start a new Livepeer instance
2. Use Livepeer CLI to set Livepeer redirect (this part is not implemented)
3. After some time, stop the old Livepeer istance

This solution sounds good at first, but I think that even with this solution a user needs some proxy, or domain name. Otherwise, they'd need to change the on-chain `Service URI` with every update/upgrade.

**4. Further work**

There are some related topics that we can cover and prioritize separately
- Prevent losing any segment in case of `MemorySession`
- Deployment in Kubernetes
- High Availability (running multiple Os behind a LB)
- [Livepeer Maintanance Mode](https://github.com/livepeer/go-livepeer/issues/2058)

**Comment by hthillman** [comment]
A few thoughts/questions:

1. Would building the redirect mechanism directly into Livepeer solve the loss of any sessions that are currently being transcoded? Or will that need to be addressed separately no matter what?
2. It sounds like no matter what this has to be set up with a proxy, correct? and the alternate solution just makes it easier / ports the work into the node itself

I am generally in favor of hiding as much complexity as we can by moving it inside the node itself. There is a wide variety of skillsets among Os, and if we can obscure the challenges behind a CLI interface that would be good.

Re: 4. Further Work, thank you for identifying these! I agree that we shouldn't discuss Kubernetes for now but I think it would be great to address in the future. Could you create stories for each of these? If we decide to tackle any of them as part of this work, we can reference them as necessary

curious for @yondonfu's thoughts

**Comment by leszko** [comment]
Thanks for the comments Hunter!

> 1. Would building the redirect mechanism directly into Livepeer solve the loss of any sessions that are currently being transcoded? Or will that need to be addressed separately no matter what?

> 2. It sounds like no matter what this has to be set up with a proxy, correct? and the alternate solution just makes it easier / ports the work into the node itself

I actually think the redirection is problematic on its own because I'm not sure what will be the address we redirect to. I can think of two cases:
- Other on-chain registered orchestrator => This could work fine, we could ask an orchestrator to gracefully shutdown (finish its work and then send redirection), but:
  - then the requirement is to have another on-chain registered orchestrator
  - should B listen to such redirection? How do we know that the new orchestrator has the same latency, stake, etc.?
- Orchestrator not registered on-chain => this opens a lot of issues, like:
  - B sometimes connect to O from on-chain, sometimes some other addresses
  - New O needs to have a public address available, which may not always be desirable
  - What to do when there is a session refresh in B? Use only B from on-chain or also these ones from redirections?

So, my thinking is that we should rather always recommend using proxy and think about how to solve all the issues we have there. It's also consistent with how all works today. Like Kubernetes, you can only access apps (pods) via proxy (services).

**Comment by leszko** [comment]
# TL;DR

There are two alternative solutions to the problem of updating/upgrading Livepeer without losing streams. Here's a comparison with Pros and Cons. Both solutions require some additional implementation.

|  | Pros | Cons |
|--|-------------------|-------------------------|
|Solution 1: Proxy | <ul><li> Consistent with how everything works nowadays<ul><li>work out of the box in Kubernetes, ECS, Fargate, etc.</li></ul></li><li>First step to achieve high availability<ul><li>proxy can automatically detect Orchestrator failures and make a switch</li></ul></li><li>First step to using multiple Orchestrators behind a load balancer</li><li>Relatively simple implementation</li><li>We promote a configuration in which an Orchestrator is not a single point of failure</li></ul> | <ul><li>Requires user to set up a proxy<ul><li>it may be as simple as running one cmd, but still it's an additional effort</li></ul></li><li>(only for TCP Proxy) Transcoding a segment may need to be retried<ul><li>it causes a delay</li><li>it may cause transcoding the same segment on two Orchestrators</li></ul></li></ul> |
| Solution 2: Redirection | <ul><li>Does not require setting up a proxy to maintain the current streams during the Livepeer update</li><li>Relatively simple implementation<ul><li>at least if we don't cover all the corner cases</li></ul></li> | <ul><li>Downtime for the on-chain Orchestrators<ul><li>at some point we remove the orchestrator which is directly registered on-chain</li></ul></li><li>Not a standard solution<ul><li>won't play well with standard solutions like Kubernetes, etc.</li><li>we may encounter some additional issues in the future</li></ul></li><li>A lot of nounces that need to be addressed:<ul><li>i.e. redirect loops, failure of the redirected orchestrator</li></ul></li></ul> |

In general, I'm in favor of implementing the **Proxy solution**, mainly because I think it allows the real zero-downtime and is more future-ready. I also think that even if we don't implement it right now, we will need to implement it in the future (i.e. to make Livepeer run in Kubernetes). What's more, the Redirection solution opens a lot of questions and has some corner cases that may be hard to address.

# Solution 1: Proxy

## Flow & User Interaction

It's the classic blue-green deployment, so the orchestrator always runs behind a reverse proxy. When a user needs to upgrade the orchestrator, then they first start a new orchestrator, then switch the proxy, and finally remove the old orchestrator.

### Step 1: Working Orchestrator 1

The user always needs to set up a proxy with the orchestrator. The simplest proxy is a one-line command which redirects the TCP stream from the port `8935` to `8937`.

```
$ socat TCP-LISTEN:8935,fork TCP:127.0.0.1:8937
```

<img src="https://user-images.githubusercontent.com/2834997/146006280-38c02585-5b32-426e-a295-88498b9e9b35.png" width=500/>

With NGINX, one can use the following configuration to proxy the TCP Stream.
```
events {}
stream {
  server {
    listen 8935;
    proxy_pass 127.0.0.1:8937;
  }
}
```

And then start `nginx`.
```
$ nginx -c nginx.conf 
```

### Step 2: Start Orchestrator 2

When the orchestrator config or version needs to be updated, a user starts a second instance of the orchestrator using a different port.

<img src="https://user-images.githubusercontent.com/2834997/146003165-af6f8e0e-9457-447b-a1cf-3f80cea789d8.png" width=500/>

### Step 3: Switch Proxy

When the new orchestrator is up and running, a user switches the proxy and the traffic is redirected to the new orchestrator.

<img src="https://user-images.githubusercontent.com/2834997/146003244-17184dd0-6fb6-4661-a71e-eae171d85558.png" width=500 />


In the case of NGINX it's an update in the configuration and applying a new config with this command.

```
$ nginx -s reload
```

###  Step 4: Remove Orchestrator 1

After a moment, orchestrator 1 can be removed.

<img src="https://user-images.githubusercontent.com/2834997/146003307-d9292ec3-c5af-42f0-8f43-0a6458ff0290.png" width=500 />

## TCP vs HTTP Proxy

Users can set up a proxy on 2 different levels TCP or HTTP. The benefit of TCP is its simplicity because its configuration is super simple. HTTP configuration is also simple, but since we use HTTPS, it requires setting SSL certs, etc.

### TCP Proxy

In the case of TCP, we don't need to think about any certs or security at the proxy level. Proxy is super simple, even one `socat` command is good enough. It redirects all the traffic from one port to another.

The drawback is that Livepeer is actually based on HTTP and HTTP/2 does not open a new TCP connection with each request. That means that during the proxy switch we need to break the TCP connection (potentially at a random point in time). That in turn means that one segment transcoding may need to be retried and potentially calculated twice (one by each orchestrator). However, I think we can implement it in a way that the stream is not lost by the orchestrator, it's just one segment that is retried.

### HTTP Proxy Implementation

In the case of HTTP, the proxy needs to be configured with SSL certificates, which might make the configuration slightly more difficult. Still not rocket science. [Here](https://docs.nginx.com/nginx/admin-guide/security-controls/securing-http-traffic-upstream/) is an example with NGINX.

The good thing is that we can implement it in a really zero-downtime manner with no retries or the double transcoding issue. Just at some point, a segment is redirected to the new orchestrator.

## Implementation changes

I think if we decide to proceed with the proxy solution, then it makes sense to make Livepeer work with both TCP and HTTP. Then, we would need to make the following implementation changes.

### (TCP Proxy) Fix retries and setting up new sessions

Currently, if a TCP is suddenly switched from one orchestrator to another, it fails with the `invalid auth token` error.

```
E1213 11:47:07.063909   72086 segment_rpc.go:88] Could not verify segment creds err=invalid auth token
```

This error is not retried from Broadcaster, so from the Orchestrator perspective, the stream is lost. We need to check this and other errors which can occur and retry them from Broadcaster. As a result, in case of an error, the transcoding of a segment is retried and processed on the new orchestrator.

### (HTTP Proxy) Fix HTTP Download

Currently, transcoding a segment does not happen in one HTTP request, but it's first transcoded and then downloaded using separate HTTP calls. It may happen that the HTTP Proxy is switched in between, which may result in HTTP Download calls executed on a different orchestrator than the transcoding HTTP Call.

There are two solutions to this issue:
- Make transcoding always in one HTTP request
- Implement/Use an ObjectStore which is shared between different orchestrators (i.e. filesystem, distributed cache)

# Solution 2: Redirection

## Flow & User Interaction

We currently return [OrchstratorInfo](https://github.com/livepeer/go-livepeer/blob/4fceaff0fce6bed65dae12428006408dc5064e15/net/lp_rpc.pb.go#L476) struct during each segment transcoding. It's possible to use one of its field (`Transcoder`) to notify Broadcaster that it should send the next segment to a different address.

### Step 1: Working Orchestrator 1

Broadcaster uses Orchestrator directly with its `host:ip` address.

<img src="https://user-images.githubusercontent.com/2834997/146004364-add4500c-5d74-478f-b0d0-d72bb8796a03.png" width=300/>

### Step 2: Start Orchestrator 2

When a user needs to update the Livepeer configuration, they start a new Orchestrator instance.

<img src="https://user-images.githubusercontent.com/2834997/146004479-1d3d0426-8fcd-4131-b0db-d6d6f20fbea4.png" width=300/>

### Step 3: Set up redirection

When Orchestrator 2 is ready, the user needs to update Orchestrator 1 with the information about the redirection.

<img src="https://user-images.githubusercontent.com/2834997/146004563-b2f8242f-6fa3-42e9-9404-b96a9abafa5e.png" width=400 />

### Step 4: Return redirection

During the next segment transcoding Orchestrator 1 returns information about the redirection to Orchestrator 2.

<img src="https://user-images.githubusercontent.com/2834997/146004617-3c38a0ba-0c0d-40b7-a058-1feac000dfa2.png" width=400/>

### Step 5: Working with Orchestrator 2

With the next segment, Broadcaster sends the data to Orchestrator 2 instead of Orchestrator 1.

<img src="https://user-images.githubusercontent.com/2834997/146004681-7d60059a-740f-4a65-9602-cf603c589f69.png" width=300 />

### Step 6: Remove Orchestrator 1

We can remove Orchestrator 1. **Note, however, that if Orchestrator 1 is directly registered for the on-chain discovery, then removing Orchestrator 1 means downtime**.

<img src="https://user-images.githubusercontent.com/2834997/146004729-cd12918d-46e0-4b2d-b080-c4b3f8c44d9e.png" width=300/>

## Comments

The new orchestrator can be an orchestrator which is registered on-chain or not registered on-chain. It will most probably always be an orchestrator under the same organization, but not registered on-chain.

There are a few issues and open questions to address:
- If the O2 is not registered on-chain, what happens if there is an on-chain discovery refresh and O1 is already removed?
- How to redirect the traffic back to the on-chain orchestrator O1? We'll probably need to add more steps to run again O1 and redirect the traffic back to O1.
- From the new Broadcaster, we don't provide zero-downtime, because there is a point in time when the on-chain orchestrator O1 is down
- How many redirects should we allow? What if there is a redirect loop?

## Implementation

We would need to implement a new endpoint in Livepeer and a new option in Livepeer CLI to allow this interaction. Then, when the redirection is set in the given orchestrator, it sends this data in the `OrchestratorInfo` for each transcoded segment.

**Comment by hthillman** [comment]
re: HTTP proxy, what is the relative difficulty of these tasks? 
- Make transcoding always in one HTTP request
- Implement/Use an ObjectStore which is shared between different orchestrators (i.e. filesystem, distributed cache)

**Comment by leszko** [comment]
> re: HTTP proxy, what is the relative difficulty of these tasks?

I think that these tasks are similar in size, I'd estimate each of them for 1-2 weeks of work. Probably 2 weeks, since everything takes more time than expected. To choose which one to implement I'd rather think about which change is more needed for the Livepeer roadmap. Maybe we should even implement both...

> * Make transcoding always in one HTTP request

The change looks very simple, but we change the B<>O protocol here, so need to be careful about the backward compatibility. The good thing about this change is that IMO we will improve the B<>O protocol and currently it's mainly livepeer.com that operates Bs, so it's simpler to make this change sooner than later.

> * Implement/Use an ObjectStore which is shared between different orchestrators (i.e. filesystem, distributed cache)

This is a bigger change, but more "clean", in a sense that we don't touch any current implementation, we just add one more implementation of the `OOSession` abstraction. I think that @yondonfu actually mentioned that the filesystem ObjectStore implementation was somewhere on the radar, but we never had time to implement it.

**Comment by leszko** [comment]
I had a very good conversation with @victorges and it's probably worth trying to make our HTTP protocol completely synchronous because now it's neither sync nor async.

So, I leaning toward trying to change the HTTP protocol and trying if it plays well with the nginx https reverse proxy.

**Comment by yondonfu** [comment]
Still digesting everything, but leaving some initial comments:

At the moment, I think focusing on making it easy for an operator run one or many Os behind a proxy makes sense as that builds off lots of existing best practices and knowledge around HA and we can make doing this more accessible via guides, ensuring that go-livepeer is easy to integrate in these setups and later on there could be additional tooling that automates much of this (i.e. `docker-compose` config file, sample Ansible playbooks, Helm charts for k8s, etc.).

Thanks for highlighting the changes in go-livepeer that would be required for HA proxy setups!

Regarding the invalid auth token error:

I think this will be a problem with an HTTP proxy as well. At the moment, an O sends B an auth token in the OrchestratorInfo message in response to a GetOrchestrator request during discovery. This auth token is then included in all segment requests from B -> O. The auth token is [generated](https://github.com/livepeer/go-livepeer/blob/970d0aa7892cd9c30aaf069a7ee8e72d8fa64111/core/orchestrator.go#L342) (using a HMAC construction) using a [secret](https://github.com/livepeer/go-livepeer/blob/970d0aa7892cd9c30aaf069a7ee8e72d8fa64111/core/orchestrator.go#L375) created on startup. So, if B switches to a different O that is behind the proxy, since the O will have a different secret, the auth token validation will fail when a new segment is received (unless the B sends a GetOrchestrator request to the new O first to get a fresh auth token, but this wouldn't happen during the redirection process with the proxy). A way to address this might be to allow the same secret to be passed to both Os so that the new O can validate the auth token included in a segment request that was originally sent to B by the old O.

Regarding the HTTP downloads:

I suspect on-disk local storage may be the easiest option for an operator to address this so they don't have to deal with cloud storage.

My main concern with this is the impact on latency if O needs to read the data from disk before B can download the results. This may or may not be that big of deal, but just flagging for further consideration. A few past conversations here for additional context and additional reasons why on-disk local storage might be useful:

https://github.com/livepeer/go-livepeer/issues/867
https://github.com/livepeer/go-livepeer/issues/1405
https://github.com/livepeer/go-livepeer/issues/846#issuecomment-489236016

I wonder if there is a temp workaround to this if the proxy can try the forwarded HTTP download request on the new O, fail and fallback to the old O. The operator would just keep the old O up for a period of time to allow remaining requests to be processed (should probably do this anyway).

That's it for now, but can follow up separately with more thoughts.

**Comment by leszko** [comment]
Thanks for the comment @yondonfu 

> Regarding the invalid auth token error:
> 
> I think this will be a problem with an HTTP proxy as well. At the moment, an O sends B an auth token in the OrchestratorInfo message in response to a GetOrchestrator request during discovery. This auth token is then included in all segment requests from B -> O. The auth token is [generated](https://github.com/livepeer/go-livepeer/blob/970d0aa7892cd9c30aaf069a7ee8e72d8fa64111/core/orchestrator.go#L342) (using a HMAC construction) using a [secret](https://github.com/livepeer/go-livepeer/blob/970d0aa7892cd9c30aaf069a7ee8e72d8fa64111/core/orchestrator.go#L375) created on startup. So, if B switches to a different O that is behind the proxy, since the O will have a different secret, the auth token validation will fail when a new segment is received (unless the B sends a GetOrchestrator request to the new O first to get a fresh auth token, but this wouldn't happen during the redirection process with the proxy). A way to address this might be to allow the same secret to be passed to both Os so that the new O can validate the auth token included in a segment request that was originally sent to B by the old O.

Yes, let me dig into this. I think your initial thought that maybe the auth secret can be shared across different Os may be the solution. Let me check it.

> Regarding the HTTP downloads:
> 
> I suspect on-disk local storage may be the easiest option for an operator to address this so they don't have to deal with cloud storage.
> 
> My main concern with this is the impact on latency if O needs to read the data from disk before B can download the results. This may or may not be that big of deal, but just flagging for further consideration. A few past conversations here for additional context and additional reasons why on-disk local storage might be useful:
> 
> #867 #1405 [#846 (comment)](https://github.com/livepeer/go-livepeer/issues/846#issuecomment-489236016)

Ok, let me check it. Maybe we'll also need some PoC to check the latency in case of the filesystem storage. Another option is a distributed cache, like Redis.

> I wonder if there is a temp workaround to this if the proxy can try the forwarded HTTP download request on the new O, fail and fallback to the old O. The operator would just keep the old O up for a period of time to allow remaining requests to be processed (should probably do this anyway).

This would need to be configured at the proxy level, but maybe it's possible 🤔

**Comment by leszko** [comment]
I dug a little deeper and did some testing/reading. Here's what we can do to address the issues around Auth Token and Download segments. Note that I didn't manage to set up the HTTPS proxy yet, so we may see some additional issues.

### Issue 1: Auth Token

One option is to make orchestrators share the same `orch.secret` (which is used to create the auth token). Currently, this secret is always [randomly generated during the startup](https://github.com/livepeer/go-livepeer/blob/763cc0150ebf47299d32f41cb3103d928c9d771e/core/orchestrator.go#L375), but we could add a flag `authTokenSecret` and therefore let the orchestrators share the same secret.

An alternative approach would be to keep the logic as it is, but when the broadcaster receives the `invalid auth token` error, it refreshes the session and retries the segment. This way, the orchestrator does not lose the stream, but the segment transcoding is **slightly delayed and may be transcoded twice**.

### Issue 2: Download segement

Since downloading the transcoded segment is performed in a separate HTTP call, it's possible that the proxy switch happens in between.

One option two solve it is to configure the fallback in the proxy as suggested by @yondonfu . According to [source 1](https://stackoverflow.com/questions/49868375/fallback-nginx-location) and [source 2](https://serverfault.com/questions/443090/how-to-configure-on-nginx-fallback-default-website), it's possible with nginx, though I haven't tried it yet.

A second option is to implement filesystem-based (or Redis-based) storage.

A third option is to change the B<>O protocol to return the transcoded data in the same HTTP call.

A fourth option is to make a retry of the segment transcoding when the download address is not found. This may result in a double-transcoding and in the delay.

### Comments

I think the simplest and the most bullet-proof solution is to add retries, but I don't know how big is the issue of a potential additional delay and possible double transcoding. I guess the delay shouldn't be such an issue, since it happens rarely and it's just for one segment, just one retry. The implications of transcoding the same segment twice by the same orchestrator should also not be a big issue. Any thoughts here? CC: @yondonfu 

Another nice solution is to share the auth token secret between orchestrators and add the proxy fallback for the segment download. It means a slightly more difficult proxy configuration for the user, but we can add a good guide on that.

**Comment by darkdarkdragon** [comment]
@leszko @yondonfu 
1. Right now we have segmented workflow - but we want to move to fully streamed one. It yet to be determined how it will look, but probably TCP or UDP connection will be constantly open between B and O. That means that proxy solution doesn't reall work for that case.

2. We also have another problem - slow session initialisation (on Nvidia cards). So, if O has some level of load and all that load being switched to another instance all at once - O will slow down significantly, and probaly will loose all the streams. In that case it could just as well just restart O instance without any fancy configuration - if it will loose streams in any case.
I think we will require to write some custom Zero-Downtime Deployment solution that will gradually move streams from old instance to new one (including new fully-streamed streams).

**Comment by leszko** [comment]
Thanks for the comment @darkdarkdragon Both are valid points!
 
> 1. Right now we have segmented workflow - but we want to move to fully streamed one. It yet to be determined how it will look, but probably TCP or UDP connection will be constantly open between B and O. That means that proxy solution doesn't reall work for that case.

Yes, that's true. However, since we still don't know how this "fully streaming pipeline" will work, I'd focus on what we have right now and short-term goals. Especially that e.g. implementing retries + guide is not a lot of work and may scratch the itch for most orchestrator operators.

Saying that I 100% agree that while designing the fully streaming pipeline we need to think about the orchestrator updates (and maybe loadbalancing + multi-replica orchestrators).

> 2. We also have another problem - slow session initialisation (on Nvidia cards). So, if O has some level of load and all that load being switched to another instance all at once - O will slow down significantly, and probaly will loose all the streams. In that case it could just as well just restart O instance without any fancy configuration - if it will loose streams in any case.
>    I think we will require to write some custom Zero-Downtime Deployment solution that will gradually move streams from old instance to new one (including new fully-streamed streams).

That's interesting, I didn't know about it.

1. Do we have any test environment to check what that slowdown? Or maybe we have some data around it? 
2. I guess most orchestrators use NVIDIA, so that this slowdown affects most of them, right?

If the slowdown is very significant (retries won't help) and affects most operators, then we may need to implement the solution with
- shared storage (or sync HTTP req)
- move the traffic gradually (using weighted load balancing) instead of the simple proxy switch

**Comment by darkdarkdragon** [comment]
@leszko 
> Do we have any test environment to check what that slowdown? Or maybe we have some data around it?

No recorded data. I think, nowadays, we have GPU dedicated for development somewhere, but I haven't used it (when I needed GPU for development I've just used VM in GCP).

> I guess most orchestrators use NVIDIA, so that this slowdown affects most of them, right?

Yes

> If the slowdown is very significant (retries won't help) and affects most operators

Slowdown is significant (when I was doing benchmarks, if I've tried to create too many sessions at once, session creation slowed down to 30s and more) - but I don't know how much it will affect our real public Os with amount of load they do have. Probably we should design some test to check what load will cause problems?

> shared storage (or sync HTTP req)

I really don't like shared storage option - better switch to sync HTTP req

**Comment by leszko** [comment]
Thanks @darkdarkdragon for the next comments. 

This task gets more and more nuanced and I think we should eat it one bite at a time. I suggest we do as follows:
1. Implement retries (seems like a low hanging fruit that scratches the itch for CPUs and GPUs with not many concurrent sessions)
2. Write a guide on how to set up an orchestrator behind a proxy and how to proceed with the orchestrator update
3. Think about how to proceed with the GPU slowdown (my first thought is that we should describe how to make the rolling session switch using the proxy and weighted load balancing, but we can consider other solutions).
4. About the streaming pipeline, I'd just take the zero-downtime assumption while designing the streaming pipeline, I don't think we should focus on it right now.

Wdyt? CC: @yondonfu

**Comment by darkdarkdragon** [comment]
@leszko 

> Implement retries (seems like a low hanging fruit that scratches the itch for CPUs and GPUs with not many concurrent sessions)


I don't think we should implement retries. If you want, create separate issue for it where we can discuss it.

> Think about how to proceed with the GPU slowdown (my first thought is that we should describe how to make the rolling session switch using the proxy and weighted load balancing, but we can consider other solutions

There is possibly that this is not a problem for a real O setups. I think it would be good to measure it first.

**Comment by leszko** [comment]
After many discussions here and also during the Technical Office Hours, we decided that we'll implement the Zero-Downtime Deployment feature during (or after) the work related to the fully streaming pipeline. I've created the related feature spec describing all the points we discussed https://github.com/livepeer/go-livepeer/issues/2169

Here are the points why we don't implement it right now:
- HTTP segment protocol we discuss right now will most probably change
- We don't see any low hanging fruit taking into consideration the NVIDIA Session Initialisation Slowdown 

Therefore, I'm closing this GH Issue.
